### DISCUSSION TITLE
Keeping the filter active after selection in multiselect ComboBox.

### DISCUSSION THREAD

**Discussion by MartynasStrazdas** [iTwin/iTwinUI#2276] [Q&A] [answered]
Hi, I was wondering if it was possible to keep the filter active in MultiSelect `ComboBox` after an option was selected? 

Screenshot for reference (Filter text is not being applied after selection):
![image](https://github.com/user-attachments/assets/d2d6a849-941b-4ddb-9648-b670df168d5d)

**Comment by mayank99** [comment]
I think this could be achieved using controlled input state (i.e. `inputProps.value`), but it's not advisable, because:
- It could conflict with the ComboBox's internal state.
- It could also lead to a confusing user experience.

Could you share more details about your use-case that requires such functionality?

**Reply by MartynasStrazdas** [reply]
We want to limit the available options in the ComboBox to a maximum of X items. One proposed solution was to load all options into the ComboBox and apply a custom filter to limit the displayed items to the first X. However, in this approach, selecting an option reveals all items, instead of maintaining the X-item limit.

While the filter’s purpose isn’t to permanently limit the number of options, this seems like the closest built in option in my eyes. My goal was to avoid pre-limiting the ComboBox options because that would require constantly tracking and including the selected items and filtering the values before providing them.

P.S. we don't have all the options from the start and load them when we need them.

**Comment by grigasp** [comment]
@mayank99, don't you think it's a bug to show unfiltered items while the filter text is still there? Have a look at this capture from your demo site:
![2024-10-04_13-22-07](https://github.com/user-attachments/assets/56cac0f7-76a3-405b-9bb3-6d10a3578aa7)

IMO, after selecting a value, either filter text should be cleared, or the options should be kept filtered. My preference would be the latter.

**Reply by mayank99** [reply]
Oh yes, this does look like a bug. The text should definitely be cleared (based on the current filter behavior). Thanks for reporting this; it wasn't clear from the original post that the text is being left over _by default_.

**Reply by grigasp** [reply]
> The text should definitely be cleared (based on the current filter behavior)

But wouldn't it be more intuitive if the filter would stay as-is, including the filtered options list? What if I want to select multiple filtered items - do I have to re-enter the filter for each of them?..

**Reply by mayank99** [reply, marked as answer]
That's a good point too. We will discuss this in more detail internally and take action appropriately. It's possible that we'll need a new prop to allow switching between the two behaviors.

**Reply by r100-stack** [reply]
> Oh yes, this does look like a bug. The text should definitely be cleared (based on the current filter behavior).

This bug has now been fixed in [3.15.4](https://github.com/iTwin/iTwinUI/releases/tag/%40itwin%2Fitwinui-react%403.15.4) 🚀 i.e. `ComboBox` with `multiple={true}` properly clears its input value whenever the filter is cleared. e.g. when an option is toggled or the combobox is unfocused.

> We will discuss this in more detail internally and take action appropriately. It's possible that we'll need a new prop to allow switching between the two behaviors.

The ability to support both behaviors is still a WIP (#2294).

**Reply by r100-stack** [reply]
> The ability to support both behaviors is still a WIP (#2294).

The ability to support both behaviors is now released in [3.16.0](https://github.com/iTwin/iTwinUI/releases/tag/%40itwin%2Fitwinui-react%403.16.0) 🚀 The behavior is controlled by the new `clearFilterOnOptionToggle` prop ([docs)](https://itwinui.bentley.com/docs/combobox#multiselect).
